include/simics/dmllib.h

```diff
@@ -2560,12 +2593,144 @@ UNUSED static void _free_table(ht_int_table_t *table) {
         ht_clear_int_table(table, true);
 }
 
-UNUSED static void _memoized_recursion(const char *name) {
+UNUSED static void _signal_critical_error(const char *restrict format, ...) {
+
+    va_list va;
+    va_start(va, format);
     char msg[512];
-    snprintf(msg, 512,
-             "Recursive call to memoized method %s. "
-             "This is considered undefined behavior.", name);
+    vsnprintf(msg, 512, format, va);
+    va_end(va);
     VT_critical_error(msg, msg);
 }
 
+UNUSED static void _memoized_recursion(const char *name) {
+    _signal_critical_error("Recursive call to memoized method %s. "
+                           "This is considered undefined behavior.", name);
+}
+
+typedef struct {
+    bool valid;
+    bool is_bank;
+    bool is_array;
+    // uint32 array_size;
+    ptrdiff_t port_obj_offset;
+    const char *portname;
+} _attr_parent_obj_proxy_info_t;
+
+typedef struct {
+    const char *name;
+    const char *type;
+    const char *doc;
+    conf_class_t *parent_obj_class;
+    _attr_parent_obj_proxy_info_t proxy_info;
+    attr_attr_t flags;
+    uint32 start_dim;
+    bool readable;
+    bool writable;
+    bool generated;
+    bool allow_cutoff;
+} _dml_attr_conf_info_t;
+
+typedef struct {
+    const _id_info_t *id_info;
+    void *vtable;
+    uint32 num;
+    bool allow_cutoff;
+} _dml_attr_getset_info_t;
+
+UNUSED static void _DML_register_attributes(
+    conf_class_t *dev_class, const _id_info_t *id_info_array,
+    const _vtable_list_t *attribute_vtables, _each_in_t sequence,
+    _dml_attr_conf_info_t (*get_attribute_info)(_traitref_t),
+    attr_value_t (*get_attr)(void *, conf_object_t *, attr_value_t *),
+    set_error_t (*set_attr)(void *, conf_object_t *, attr_value_t *,
+                            attr_value_t *),
+    attr_value_t (*get_portobj_attr)(void *, conf_object_t *, attr_value_t *),
+    set_error_t (*set_portobj_attr)(void *, conf_object_t *, attr_value_t *,
+                            attr_value_t *)) {
+    for (uint32 i = 0; i < sequence.num; ++i) {
+        _vtable_list_t list = attribute_vtables[i];
+        uint64 num = list.num / sequence.array_size;
+        uint64 start = num * sequence.array_idx;
+        _traitref_t traitref = { list.vtable,
+                                 { .id = list.id, .encoded_index = start }};
+        _dml_attr_conf_info_t attr_info = get_attribute_info(traitref);
+        if (!attr_info.generated) {
```

Comment:
Confused me on my first read, past tense indicates it means "operation succeeded" rather than "please do it". Also I'd say we register stuff rather than generate. I suggest `skip`.

Reply:
`generated` now replaced with `should_be_generated`. I prefer that to `skip`.